seq_generator.py

```diff
@@ -84,13 +87,26 @@ def inc_repeats():
     result["symbols"] = grammar.symbols
     return result
 
-def create_dataset(min_length, max_length, num_samples, loc_grammar, loc_dataset):
+def create_dataset(min_length, max_train_length, max_test_length, num_train_samples, num_id_samples,
+                   num_ood_samples, loc_grammar, loc_dataset):
     grammar = ruleset.Ruleset()
     grammar.load(loc_grammar)
 
-    r = generate_data(min_length, max_length, num_samples, grammar)
+    train_data = generate_data(min_length, max_train_length, num_train_samples, grammar, "train")
     with open(loc_dataset, "w") as f:
-        json.dump(r, f, indent=4)
+        json.dump(train_data, f, indent=4)
+
+    non_terminals = list(grammar.rules.keys())
+
+    # omit unproductive rules from test sets
+    for nt in non_terminals:
+        for rule in grammar.rules[nt]:
+            if (not rule[1]):
```

Comment:
I am not 100 % sure if this is correct. You mark a rule (e.g. rule 3) as used if it ever appeared during generation right? However the way we describe our dataset rule 3 might be able to produce [4, 5], "a" or [1, 2]. If now rule 3 appears during generation and the rule 3 -> [4, 5] is used then the way this is written now would simply say that rule 3 has been used, but actually the rule 3 -> "a" or 3 -> [1, 2] might have never been used.

Reply:
No, the way I mark rules is that I extend every "subrule" into a list whose second entry corresponds to the used flag. In this stage, I delete every subrule whose second entry is False. Thank You for making sure though.
![image](https://github.com/jnice-81/DLCykParser/assets/48862915/7deacae7-b221-49a3-ada6-fad332217d6f)

Reply:
Maybe I can rename the variable such that this is clearer?

Reply:
Ahh I see. Naa guess this is fine then